addons/google_gmail/models/google_gmail_mixin.py

```diff
@@ -0,0 +1,104 @@
+            values['google_gmail_refresh_token'] = self.env['google.service'].generate_refresh_token('gmail', values['google_gmail_authorization_code'])
+
+        return super(GoogleGmailMixin, self).write(values)
+
+    @api.depends('google_gmail_authorization_code')
+    def _compute_gmail_uri(self):
+        google_gmail_uri = self.env['google.service']._get_google_token_uri('gmail', scope=self._get_google_scope())
+        for server in self:
+            server.google_gmail_uri = google_gmail_uri
+
+    @api.model
+    def _get_access_token(self, refresh_token):
+        Config = self.env['ir.config_parameter'].sudo()
+        user_is_admin = self.env.user._is_admin()
+
+        google_gmail_client_id = Config.get_param('google_gmail_client_id')
+        google_gmail_client_secret = Config.get_param('google_gmail_client_secret')
+
+        if not google_gmail_client_id or not google_gmail_client_secret:
+            raise UserError(
+                _('Google Gmail is not yet configured.')
+                if user_is_admin else
+                _('Google Gmail is not yet configured. Please contact your administrator.')
+            )
+
+        if not refresh_token:
+            raise UserError(
+                _('The refresh token for authentication is not set.')
+                if user_is_admin else
+                _('Google Gmail is not yet configured. Please contact your administrator.')
+            )
+
+        try:
+            result = requests.post(
+                GOOGLE_TOKEN_ENDPOINT,
+                data={
+                    'client_id': google_gmail_client_id,
+                    'client_secret': google_gmail_client_secret,
+                    'refresh_token': refresh_token,
+                    'grant_type': 'refresh_token',
+                    'scope': self._get_google_scope(),
+                },
+                headers={'Content-type': 'application/x-www-form-urlencoded'},
+                timeout=TIMEOUT,
+            )
+            result.raise_for_status()
+        except requests.HTTPError:
+            raise UserError(
+                _('Something went wrong during the token generation. Please request again an authorization code .')
+                if user_is_admin else
+                _('Google Gmail is not yet configured. Please contact your administrator.')
+            )
+
+        return result.json().get('access_token')
+
+    def _get_google_scope(self):
+        return 'https://mail.google.com/'
+
+    def _generate_oauth2_string(self, user, refresh_token):
+        """Generate a OAuth2 string which can be used for authentication.
+
+        :param user: Email address of the Gmail account to authenticate
+        :param refresh_token: Refresh token for the given Gmail account
+
+        :return: The SASL argument for the OAuth2 mechanism.
+        """
+        access_token = self._get_access_token(refresh_token)
+        return f'user={user}\1auth=Bearer {access_token}\1\1'
```

Comment:
For odoo 12.0, we still need to support python 3.5, which doesn't know about f-strings.
```suggestion
        return 'user={user}\1auth=Bearer {access_token}\1\1'.format(user=user, access_token=access_token)
```